synapse/cryotank.py

```diff
@@ -571,3 +589,644 @@ def init(self, name, conf=None, timeout=None):
+
+        self._worker.start()
+
+        def _onfini():
+            self._workq.done()
+            self._worker.join(self.MAX_WAIT_S)
+            self._dbenv.close()
+
+        ebus.onfini(_onfini)
+
+    def _onData(self, unused):
+        '''
+        Wake up the worker if he already doesn't have a reason to be awake
+        '''
+        if 0 == len(self._workq):
+            self._workq.put((None, lambda: None, None, None))
+
+    def _removeSome(self):
+        '''
+        Make some progress on removing deleted indices
+        '''
+        left = self._remove_chunk_sz
+        for iid in self._meta.deleting:
+            if not left:
+                break
+            iid_enc = _iid_en(iid)
+            with self._dbenv.begin(db=self._idxtbl, buffers=True, write=True) as txn, txn.cursor() as curs:
+                if curs.set_range(iid_enc):
+                    for k, offset_enc in curs.iternext():
+                        if k[:len(iid_enc)] != iid_enc:
+                            break
+                        if not curs.delete():
+                            raise s_exc.CorruptDatabase('delete failure')
+
+                        txn.delete(offset_enc, iid_enc, db=self._normtbl)
+                        left -= 1
+                        if not left:
+                            break
+                if not left:
+                    break
+
+            self._meta.markDeleteComplete(iid)
+
+    def _normalize_records(self, raw_records):
+        '''
+        Yields stream of normalized fields
+
+        Args:
+            raw_records(Iterable[Tuple[int, Dict[int, str]]])  generator of tuples of offset/decoded raw cryotank
+            record
+        Returns:
+            Iterable[Tuple[int, int, Union[str, int]]]: generator of tuples of offset, index ID, normalized property
+            value
+        '''
+        for offset, record in raw_records:
+            self._next_offset = offset + 1
+            dp = s_datapath.initelem(s_msgpack.un(record))
+
+            for iid, idx in ((k, v) for k, v in self._meta.indices.items() if not self._meta.asleep[k]):
+                if self._meta.progresses[iid]['nextoffset'] > offset:
+                    continue
+                try:
+                    self._meta.progresses[iid]['nextoffset'] = offset + 1
+                    for datapath in idx.datapaths:
+                        field = dp.valu(datapath)
+                        if field is None:
+                            continue
+                        # TODO : what to do with subprops?
+                        break
+                    else:
+                        logger.debug('Datapaths %s yield nothing for offset %d',
+                                     [d.path for d in idx.datapaths], offset)
+                        continue
+                    normval, _ = s_datamodel.getTypeNorm(idx.syntype, field)
+                except (s_exc.NoSuchType, s_exc.BadTypeValu):
+                    logger.debug('Norm fail')
```

Comment:
this debug message isn't super useful - we should probably chuck it or add contextual information and accept that index fails may produce a huge amount of logs

Reply:
I added the exception to the output. Yeah was on the fence about this, but it will be highly useful to debug normalization failures. Debug logging isn't normally enabled, right?

Reply:
normaly we don't run production things with debug logging.  if we have an index that is only norming values 50% of the time (which may be entirely reasonable), we'd potentially end up dumping ALOT of content into the logs if we included the valu.  Even worse, if the datapath failed to map to a value, and returns the whole record, we'd be in a worse case.  So...maybe we just rip this log message out entirely and can leave norm-debugging as an exercise to a user.  We'll already know when we have bunk records since the fail counters will increase.